rust/arrow/src/compute/kernels/sort.rs

```diff
@@ -0,0 +1,295 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Defines sort kernel for `ArrayRef`
+
+use std::cmp::Reverse;
+
+use crate::array::*;
+use crate::compute::take;
+use crate::datatypes::*;
+use crate::error::{ArrowError, Result};
+
+use TimeUnit::*;
+
+/// Sort the `ArrayRef` using `SortOptions`.
+///
+/// Performs a stable sort on values and indices, returning nulls after sorted valid values,
+/// while preserving the order of the nulls.
+///
+/// Returns and `ArrowError::ComputeError(String)` if the array type is either unsupported by `sort_to_indices` or `take`.
+pub fn sort(values: &ArrayRef, options: Option<SortOptions>) -> Result<ArrayRef> {
+    let indices = sort_to_indices(values, options)?;
+    take(values, &indices, None)
+}
+
+/// Sort elements from `ArrayRef` into an unsigned integer (`UInt32Array`) of indices
+pub fn sort_to_indices(
+    values: &ArrayRef,
+    options: Option<SortOptions>,
+) -> Result<UInt32Array> {
+    let options = options.unwrap_or(Default::default());
+    let range = values.offset()..values.len();
+    let (v, n): (Vec<usize>, Vec<usize>) =
+        range.partition(|index| values.is_valid(*index));
+    let n = n.into_iter().map(|i| i as u32).collect();
+    match values.data_type() {
+        DataType::Boolean => sort_primitive::<BooleanType>(values, v, n, &options),
+        DataType::Int8 => sort_primitive::<Int8Type>(values, v, n, &options),
+        DataType::Int16 => sort_primitive::<Int16Type>(values, v, n, &options),
+        DataType::Int32 => sort_primitive::<Int32Type>(values, v, n, &options),
+        DataType::Int64 => sort_primitive::<Int64Type>(values, v, n, &options),
+        DataType::UInt8 => sort_primitive::<UInt8Type>(values, v, n, &options),
+        DataType::UInt16 => sort_primitive::<UInt16Type>(values, v, n, &options),
+        DataType::UInt32 => sort_primitive::<UInt32Type>(values, v, n, &options),
+        DataType::UInt64 => sort_primitive::<UInt64Type>(values, v, n, &options),
+        // DataType::Float32 => sort_primitive::<Float32Type>(values, v, n, descending),
```

Comment:
could you add a comment explaining why the floating point ones are commented out?

Reply:
I think I was trigger-happy when thinking that this was ready for review. Floats can't be sorted in the default way, so I need `ordered_float` or an equivalent. I left floats commented out because I wanted to investigate this.

I have at least 2 options, depending on which you prefer;

1. leave a note and open a JIRA to support floats
2. keep this PR open for longer so I can add the float support.

I'll try the latter in the interim.

Reply:
Having a separate JIRA for sorting floats sounds good to me.

Reply:
https://issues.apache.org/jira/browse/ARROW-7924